[CSSTransition] Do not list up all properties for transition: all

Constellation opened the pull request:
#### 66eb8a0eb60e88dc4a31d1ee690e7fe132267708
<pre>
[CSSTransition] Do not list up all properties for transition: all
<a href="https://bugs.webkit.org/show_bug.cgi?id=261917">https://bugs.webkit.org/show_bug.cgi?id=261917</a>
rdar://115861030

Reviewed by Antti Koivisto.

When `transition: all ...` CSS rule is applied, we end up listing all animatable properties.
This is incredibly costly since there are so many CSS properties. And all properties are queried,
checked equality via very slow virtual function. Many hash table lookups etc.
We found that we are using massive amount of time in this transitioning in Speedometer3/NewsSite-Next / NewsSite-Nuxt.

But in almost all cases, almost all properties are not related at all.

In this patch, we do RenderStyle comparison and collect all affected CSS properties candidate which needs to be visited in
updateCSSTransitions. We can conservatively say &quot;this property needs to be visited&quot;. But if we know that this
is totally equal, then we can skip that.

We cannot use RenderStyle::diff for this purpose. It is highly tailored to rendering purpose. And a lof of
properties changes are not reported when it is unrelated to layout (but it is necessary for updateCSSTransitions).
Example is `top` / `left`. So instead, we have more tailored path collecting necessary properties in this patch.

With this patch, we always collect all possible affected properties. So we no longer need to list up all animatable properties.
It results in 0.7% progression in Speedometer3.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasAnimations const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::hasAnimations const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::updateCSSTransitions const):

Canonical link: <a href="https://commits.webkit.org/268412@main">https://commits.webkit.org/268412@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a6f45adf314540f2ca2740c10924bf7d40b2e17d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19611 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21504 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18344 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19848 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23295 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20174 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19829 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19845 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17052 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22361 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17843 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24149 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18093 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18017 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22125 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18623 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15789 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17770 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22125 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2400 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18449 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4510,6 +4510,13 @@ bool Element::hasKeyframeEffects(PseudoId pseudoId) const
+{
+    if (!hasRareData())
+        return false;
+    return elementRareData()->hasAnimations();
```
nit: could be a one liner:
`return hasRareData() && elementRareData()->hasAnimations()`

Constellation replied:
Fixed.

cdumez commented on Source/WebCore/style/Styleable.cpp:
```diff
@@ -683,13 +683,40 @@ void Styleable::updateCSSTransitions(const RenderStyle& currentStyle, const Rend
+            }
+
+            // Because we do not have keyframeEffectStack() and lastStyleChangeEventStyle(), currentStyle and newStyle will be directly used.
+            OptionSet<StyleDifferenceContextSensitiveProperty> properties { };
```
`{ }` shouldn't be needed.

Constellation replied:
Fixed.

cdumez reviewed (commented):
Not familiar enough with this code to know if this change is OK.

anttijk commented on Source/WebCore/animation/WebAnimationTypes.h:
```diff
@@ -71,6 +72,11 @@ using CSSAnimationCollection = ListHashSet<Ref<CSSAnimation>>;
+class CSSPropertiesBitSet {
+public:
+    WTF::BitSet<numCSSProperties> m_properties { };
+};
```
Does this really need a class (struct?) rather than `using CSSPropertiesBitSet = WTF::BitSet<numCSSProperties>`?

Constellation replied:
The reason of this is not adding `CSSPropertyNames.h` dependency from `RenderStyle.h`, since `numCSSProperties` is defined in that header.

anttijk replied:
numCSSProperties could be generated to a header of its own.

anttijk replied:
Please at least make this a struct and use struct style member naming (withou m_ prefix).

Constellation replied:
Thanks, I've changed it to `struct` for now.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1422,6 +1422,678 @@ bool RenderStyle::diffRequiresLayerRepaint(const RenderStyle& style, bool isComp
+{
+    // This function conservatively answsers two things,
+    //   1. Whether we can use CSSTransitionAll fast path.
+    //   2. What CSS properties we should visit in the fast path.
```
How does it answer the first question negatively? What happens then?

This really does seem like something that should be autogenerated from CSSProperties.json. It will be difficult to keep it up to date when new properties are added.

Constellation replied:
Now, this patch covers all the cases, and removed the fast path. So (1) should be removed, it is always handled in this code.

> This really does seem like something that should be autogenerated from CSSProperties.json. It will be difficult to keep it up to date when new properties are added.

It is ideal. But right now, we do not have a mechanism to generate data structures, and `RenderStyle::diff` is also not doing it. So for now, this patch is aligning the approach to `RenderStyle::diff`. When introducing a mechanism to solve `RenderStyle::diff` etc. too, this probably can be organized in the same way.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1422,6 +1422,678 @@ bool RenderStyle::diffRequiresLayerRepaint(const RenderStyle& style, bool isComp
     return false;
 }
 
+void RenderStyle::collectAffectedProperties(const RenderStyle& other, CSSPropertiesBitSet& changingProperties) const
```
Affected how? I think this is more like `collectChangedProperties` or `conservativelyCollectChangedProperties`

But I suppose this currently limited to animatable properties so `conservativelyCollectChangedAnimatableProperties` or something.

Constellation replied:
Sounds good. Changed it to `conservativelyCollectChangedAnimatableProperties`.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1422,6 +1422,678 @@ bool RenderStyle::diffRequiresLayerRepaint(const RenderStyle& style, bool isComp
+    //   1. Whether we can use CSSTransitionAll fast path.
+    //   2. What CSS properties we should visit in the fast path.
+    // (2) is conservative. We do not need to precisely check equivalence before saying "this property needs to be visisted".
+    // We can extend this function to increase coverage of this fast path by adding more properties.
```
As far as I can tell any missing property here will just fail to animate?

Constellation replied:
Yes, for `transition all` case.

anttijk commented on Source/WebCore/rendering/style/SVGRenderStyle.cpp:
```diff
@@ -232,4 +232,137 @@ bool SVGRenderStyle::changeRequiresRepaint(const SVGRenderStyle& other, bool cur
     return false;
 }
 
+void SVGRenderStyle::conservativelyCollectChangedAnimatableProperties(const SVGRenderStyle& other, CSSPropertiesBitSet& changingProperties) const
```
please add a FIXME about autogenerating all this.

Constellation replied:
Added.

Constellation committed:
[CSSTransition] Do not list up all properties for transition: all
https://bugs.webkit.org/show_bug.cgi?id=261917
rdar://115861030

Reviewed by Antti Koivisto.

When `transition: all ...` CSS rule is applied, we end up listing all animatable properties.
This is incredibly costly since there are so many CSS properties. And all properties are queried,
checked equality via very slow virtual function. Many hash table lookups etc.
We found that we are using massive amount of time in this transitioning in Speedometer3/NewsSite-Next / NewsSite-Nuxt.

But in almost all cases, almost all properties are not related at all.

In this patch, we do RenderStyle comparison and collect all affected CSS properties candidate which needs to be visited in
updateCSSTransitions. We can conservatively say "this property needs to be visited". But if we know that this
is totally equal, then we can skip that.

We cannot use RenderStyle::diff for this purpose. It is highly tailored to rendering purpose. And a lof of
properties changes are not reported when it is unrelated to layout (but it is necessary for updateCSSTransitions).
Example is `top` / `left`. So instead, we have more tailored path collecting necessary properties in this patch.

With this patch, we always collect all possible affected properties. So we no longer need to list up all animatable properties.
It results in 0.7% progression in Speedometer3.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasAnimations const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::hasAnimations const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::updateCSSTransitions const):

Canonical link: https://commits.webkit.org/268412@main
